packages/flutter/lib/src/painting/matrix_utils.dart

```diff
@@ -124,9 +123,22 @@ class MatrixUtils {
   /// This function assumes the given point has a z-coordinate of 0.0. The
   /// z-coordinate of the result is ignored.
   static Offset transformPoint(Matrix4 transform, Offset point) {
-    final Vector3 position3 = Vector3(point.dx, point.dy, 0.0);
-    final Vector3 transformed3 = transform.perspectiveTransform(position3);
-    return Offset(transformed3.x, transformed3.y);
+    final Float64List storage = transform.storage;
+    final double x = point.dx;
+    final double y = point.dy;
+
+    // Directly simulate the transform of the vector (x, y, 0, 1),
+    // dropping the resulting Z coordinate, and normalizing only
+    // if needed.
+
+    final double rx = storage[0] * x + storage[4] * y + storage[12];
+    final double ry = storage[1] * x + storage[5] * y + storage[13];
+    final double rw = storage[3] * x + storage[7] * y + storage[15];
```

Comment:
OK, that explains some of the backwards code I saw in the vector_math package.

(I'm used to Java which would have handled this differently.)

Reply:
I'm not sure this makes sense, but I implemented both of these "range check" proposals and they slowed it down on both of my test devices.  I got the following results:

Moto E2:
```
I/flutter ( 5449): MatrixUtils.transformRectPerspective: 1038.7 ns per iteration
I/flutter ( 5449): MatrixUtils.transformRectAffine: 663.2 ns per iteration
I/flutter ( 5449): MatrixUtils.transformPointPerspective: 514.4 ns per iteration
I/flutter ( 5449): MatrixUtils.transformPointAffine: 489.8 ns per iteration
```

Samsung Note 9:
```
I/flutter (14052): MatrixUtils.transformRectPerspective: 86.3 ns per iteration
I/flutter (14052): MatrixUtils.transformRectAffine: 56.2 ns per iteration
I/flutter (14052): MatrixUtils.transformPointPerspective: 44.1 ns per iteration
I/flutter (14052): MatrixUtils.transformPointAffine: 41.6 ns per iteration
```

I ran it multiple times to see if they were flukes, but they never were as fast as the runs I did on the last commit. :(

Reply:
To be specific, here is the actual change I made to put a reference to storage[15] first in both methods:

```
diff --git a/packages/flutter/lib/src/painting/matrix_utils.dart b/packages/flutter/lib/src/painting/matrix_utils.dart
index c6da1c550..55b5ab742 100644
--- a/packages/flutter/lib/src/painting/matrix_utils.dart
+++ b/packages/flutter/lib/src/painting/matrix_utils.dart
@@ -131,9 +131,9 @@ class MatrixUtils {
     // dropping the resulting Z coordinate, and normalizing only
     // if needed.
 
-    final double rx = storage[0] * x + storage[4] * y + storage[12];
-    final double ry = storage[1] * x + storage[5] * y + storage[13];
-    final double rw = storage[3] * x + storage[7] * y + storage[15];
+    final double rw = storage[15] + storage[3] * x + storage[7] * y;
+    final double rx = storage[12] + storage[0] * x + storage[4] * y;
+    final double ry = storage[13] + storage[1] * x + storage[5] * y;
     if (rw == 1.0) {
       return Offset(rx, ry);
     } else {
@@ -301,6 +301,10 @@ class MatrixUtils {
     // optimized perspective = 12 mul + 39 add/sub/compare
     // optimized affine      =  8 mul + 17 add/sub/compare
 
+    final bool isAffine = storage[15] == 1.0 &&
+                          storage[ 3] == 0.0 &&
+                          storage[ 7] == 0.0;
+
     final double wx = storage[0] * w;
     final double hx = storage[4] * h;
     final double rx = storage[0] * x + storage[4] * y + storage[12];
@@ -309,7 +313,7 @@ class MatrixUtils {
     final double hy = storage[5] * h;
     final double ry = storage[1] * x + storage[5] * y + storage[13];
 
-    if (storage[3] == 0.0 && storage[7] == 0.0 && storage[15] == 1.0) {
+    if (isAffine) {
       double left  = rx;
       double right = rx;
       if (wx < 0) {
```

Reply:
I tried this again with back to back runs on both devices with and without that patch and the benchmark modified to use a lot more runs and the timings I got with the proposed commit were faster than the timings I got with the additional patch in the previous comment in most cases. The cases where changing the order showed a win were isolated and slim. The cases where leaving it as submitted showed a win were more significant and more numerous and the averages were in favor of the fix as submitted.  This is really in the noise in the end so I'm going to leave it as I've already submitted it.

https://docs.google.com/spreadsheets/d/1FhOgYeEfpY52iUQiUGvx3xWi4ROd2HuGyIvsy1lgqms/edit?usp=sharing

Reply:
/cc @mraleph, who could shed some light on what's going on behind the scenes. I didn't expect a huge improvement from eliminating range checks, but I certainly didn't expect a slow down 😕 

**tl;dr** an optimization we tried is to read the last index out of a `Float64List` first so that the compiler only range checks once and lets smaller indices bypass the range check, i.e.:

```
final a = ...;
// The following should be faster
a[15]; a[10]; a[5]; a[0];
// than this
a[0]; a[5]; a[10]; a[15];
```

Because in the first case if `a[15]` doesn't throw a range error, the following statements cannot possibly throw due to smaller indices, so a range check isn't necessary. But @flar is observing a _slow down_.

Reply:
The slow down most likely originated from creating boolean and the dispatching on the boolean `if (isAffine) {` part of the change. Materialising boolean and dispatching on it later is more expensive than the chain of comparisons. 

... but I also discovered that we have a regression in our bounds check elimination - we no longer would eliminate bounds checks in a sequence like what @yjbanov written. We used to be able to do that, but that got broken - and we did not have tests for it. I filed https://github.com/dart-lang/sdk/issues/37687 to address that. 

[Benefits of eliminating bounds checks are pretty minimal though especially on the code where the performance is dominated by memory traffic and floating point math]

Reply:
A question on bounds check elimination...

If the code had `if (a.length < 16) return null;` towards the top (with the understanding that it would never happen so returning null is mostly vestigial), would the compiler be able to eliminate all bounds checking for constant integer indices < 16?

Reply:
The answer to that question is complicated... Theoretically `if (a.length < 16) return null;` should have the same effect as just loading `a[15]`... And in JIT it would, but in practice it is broken right now too :-/

Reply:
(`if (a.length < 16) throw 'impossible';` would be better because it does not polute type information about the return value).